Prevent overcompilation in mixed projects without pipelining

eatkins opened the pull request:
The latest zinc will overcompile mixed projects when scala sources have changed but no java sources have. Fixes #867.

eatkins committed:
Create java class directory if missing

It is possible to invoke the java compiiler in  astate where the output
directory doens't exist. When this happens, the compiler throws an
IllegalStateException:

[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: {line 8} Command failed: 'run'
[error] Caused by: java.lang.IllegalStateException: directory not found: /var/folders/zv/2wskcxc14cn6kyj4yb53whc40000gn/T/sbt_4bd7559c/target/classes
[error]         at com.sun.tools.javac.main.Main.error(Main.java:186)
[error]         at
com.sun.tools.javac.main.Main.checkDirectory(Main.java:345)
[error]         at
com.sun.tools.javac.main.Main.processArgs(Main.java:274)
[error]         at com.sun.tools.javac.main.Main.compile(Main.java:414)
[error]         at
com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
[error]         at
com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
[error]         at
sbt.internal.inc.javac.LocalJavaCompiler.run(LocalJava.scala:327)
[error]         at
sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$12(AnalyzingJavaCompiler.scala:167)
...

eatkins committed:
Only recompile all java sources if any have changed

When any java source is invalidated, we want to recompile the entire
subproject to generate the java class signatures. For this reason, it is
necessary to invalidate _all_ of the java sources when any change is
detected. We don't however want to recompile all of the java sources if
_none_ of them have changed.

Prior to this change, in the sbt build with pipelining off, changing any
scala source in a mixed project would rebuild all of the java sources
which could trigger a cascade of invalidations that could potentially
invalidate the entire project even when none of the java sources had
actually changed. Due to this behavioral change, in
0504f70454fdbe9db2cba4003c710888f6e31e2a it was necessary to change the
iteration count in less-inter-inv-java test from 3 to 2 because the
overcompiling behavior made it only do one cycle when a single source
file was changed because all of the sources were automatically
invalidated. This commit restores the old behavior so the test can be
restored as well. It also adds a test for mixed projects to ensure that
changing a scala file does not automatically trigger recompilation of
java files.

eatkins committed:
Add scripted source files to scripted watch triggers

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/868/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins committed:
Preserve line endings in scalafmt

dwijnand commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -340,18 +340,21 @@ object Incremental {
       incremental.invalidateInitial(previous.relations, initialChanges)
 
     // If there's any compilation at all, invalidate all java sources too, so we have access to their type information.
```
What about this part?  Does this change lead to a loss of type information?

eed3si9n replied:
That comment was added in https://github.com/sbt/zinc/commit/0504f70454fdbe9db2cba4003c710888f6e31e2a (port of https://github.com/mspnf/zinc/blob/9ef2eba6f829c2afa783987f64ea4e785b678265/internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala). The type information that this makes it available is to the downstream subproject compilation, which is only relevant in pipelining.

dwijnand replied:
Ah, cool, thanks.

eatkins committed:
Always create output directories

We should always call Files.createDirectories to ensure the output
directory exists. If the directory already exists and is a directory,
Files.createDirectories is a no-op and is the same amount of io as the
Files.exists check. If the directory exists but is actually a file, then
that is a bug and compilcation is likely to fail so we should throw an
exception.

eed3si9n commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -340,17 +340,17 @@ object Incremental {
     val initialInvSources =
-      if (initialInvSources0.nonEmpty) initialInvSources0 ++ javaSources
-      else Set.empty[VirtualFileRef]
+      if (isPickleJava && initialInvSources0.nonEmpty) initialInvSources0 ++ javaSources
+      else initialInvSources0
```
👍  There's no point in invalidating Java sources unless `isPickleJava` is on.

eatkins replied:
Cool. I was about to submit a comment justifying this choice but I don't need to since you've confirmed my intuition.

eed3si9n reviewed (approved):
LGTM

eed3si9n commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -340,17 +340,17 @@ object Incremental {
       incremental.invalidateInitial(previous.relations, initialChanges)
 
     // If there's any compilation at all, invalidate all java sources too, so we have access to their type information.
```
```suggestion
    // During pipelining, if there's any compilation at all, invalidate all Java sources too, so the downstream Scala subprojects would have type information via early output (pickle jar).
```

eed3si9n committed:
Update comment about invalidating all Java sources

eatkins reviewed (commented):
👍
